feat!: `fuel-core@0.32.1` and large contract deployments (#2827)
FuelLabs/fuels-ts: apps/docs/src/guide/contracts/deploying-contracts.md

# Deploying Contracts

> Deploying contracts using the SDK is handled by the `ContractFactory`. The process involves collecting the contract artifacts, initializing the contract factory, and deploying the contract.

Reviewer: ```suggestion
To deploy a contract using the SDK, you can use the `ContractFactory`. This process involves collecting the contract artifacts, initializing the contract factory, and deploying the contract.
```

> The SDK utilizes two different deployment processes. Either simply using a single create transaction to deploy the entire contract bytecode, or, by splitting the contract into multiple chunks, deploying them as blobs (on chain data accessible to the VM) and then generating a contract from the associated blob IDs. That generated contract is then deployed as a create transaction.

Reviewer 2: ```suggestion
The SDK utilizes two different deployment processes, depending on the contract's size being less than or equal to the `contract_max_size` in the chain's consensus parameters. It either uses a single create transaction to deploy the entire contract bytecode, or it splits the contract bytecode into multiple chunks, deploys them as blobs (on chain data accessible to the VM), and then generates a contract from the associated blob IDs. That generated contract is then deployed as a create transaction.
```

The `ContractFactory` offers the following methods for the different processes:

- `deploy` for deploying contacts of any size (will automatically choose the appropriate deployment process).
- `deployContract` for deploying the entire contract bytecode in a single create transaction.
- `deployContractAsBlobs` for deploying the contract in chunks as blobs, and then deploying the contract as a create transaction.

> > **Note:** Due to the nature of blob deployments, both `deploy` and `deployContractAsBlobs` may require multiple transactions to deploy the contract.

Reviewer 2: ```suggestion
> **Note:** If the contract is deployed via blob deployments, multiple transactions will be required to deploy the contract.
```

> The SDK utilizes two different deployment processes. Either simply using a single create transaction to deploy the entire contract bytecode, or, by splitting the contract into multiple chunks, deploying them as blobs (on chain data accessible to the VM) and then generating a contract from the associated blob IDs. That generated contract is then deployed as a create transaction.
> 
> The `ContractFactory` offers the following methods for the different processes:
> 
> - `deploy` for deploying contacts of any size (will automatically choose the appropriate deployment process).
> - `deployContract` for deploying the entire contract bytecode in a single create transaction.
> - `deployContractAsBlobs` for deploying the contract in chunks as blobs, and then deploying the contract as a create transaction.
> 
> > **Note:** Due to the nature of blob deployments, both `deploy` and `deployContractAsBlobs` may require multiple transactions to deploy the contract.
> 
> The deployment process used by the `ContractFactory` is dependent on the size of the contract you are attempting to deploy. The threshold for the contract size is dictated by the chain. You can find the maximum contract size supported by the chain by querying the chains consensus parameters.
> 
> <<< @/../../docs-snippets/src/guide/contracts/deploying-contracts.test.ts#get-contract-max-size{ts:line-numbers}
> 

Reviewer 2: ```suggestion
The SDK utilizes two different deployment processes, depending on the contract's size. The threshold for the contract size is dictated by the chain size and can be queried like this:

<<< @/../../docs-snippets/src/guide/contracts/deploying-contracts.test.ts#get-contract-max-size{ts:line-numbers}

It either uses a single create transaction to deploy the entire contract bytecode, or it splits the contract bytecode into multiple chunks, deploys them as blobs (on chain data accessible to the VM), and then generates a contract from the associated blob IDs. That generated contract is then deployed as a create transaction.

The `ContractFactory` offers the following methods for the different processes:

- `deploy` for deploying contacts of any size (will automatically choose the appropriate deployment process).
- `deployContract` for deploying the entire contract bytecode in a single create transaction.
- `deployContractAsBlobs` for deploying the contract in chunks as blobs, and then deploying the contract as a create transaction.

> **Note:** Due to the nature of blob deployments, both `deploy` and `deployContractAsBlobs` may require multiple transactions to deploy the contract.
```

Reviewer 2: Note that this is just a take on the docs that flows better for me. Pick and choose what you will from it.

Author: Sorry @nedsalk I haven't missed this. Will compare today.

Author: It's nice 👍🏻

<<< @/../../docs-snippets/src/guide/contracts/deploying-contracts.test.ts#deploy{ts:line-numbers}

The `contract` instance will be returned only after calling `waitForResult` and waiting for it to resolve. To avoid blocking the rest of your code, you can attach this promise to a hook or listener that will use the contract only after it is fully deployed.

### 3. Executing a Contract Call

Now that the contract is deployed, you can interact with it by submitting a contract call:

<<< @/../../docs-snippets/src/guide/contracts/deploying-contracts.test.ts#call{ts:line-numbers}

## Deploying a Large Contract as Blobs

In the above guide we use the recommended `deploy` method. If you are working with a contract that is too large to be deployed in a single transaction, then the SDK will chunk the contract for you and submit it as blobs, to then be accessed later by a create transaction. This process is handled by the `deployContractAsBlobs` method, also available on the `ContractFactory` should you want to use that directly.

<<< @/../../docs-snippets/src/guide/contracts/deploying-contracts.test.ts#blobs{ts:line-numbers}

> In the above example, we also pass a `chunkSizeOverride` option to the deployment method. The SDK will attempt to chunk the contract to the most optimal about, however the transaction size can fluctuate and you can also be limited by request size limits against the node. By default we set an of 5%, however you can adjust this to suit your needs and ensure the transaction passes. It must be set to a value between 0 and 1.

Reviewer 3: ```suggestion
In the above example, we also pass a `chunkSizeOverride` option to the deployment method. The SDK will attempt to chunk the contract to the most optimal amount, however, the transaction size can fluctuate and you can also be limited by request size limits against the node. By default, we set a value for `chunkSizeOverride` of 0.05 (5%), however you can adjust this to suit your needs and ensure the transaction passes. It must be set to a value between 0 and 1.
```
